src/librustc_mir/tcx/mod.rs

```diff
@@ -0,0 +1,179 @@
+        self.cmp_method_ref(eq_def_id, "eq", ty)
+    }
+
+    fn partial_le(&mut self, ty: Ty<'tcx>) -> ItemRef<Self> {
+        let ord_def_id = self.tcx.lang_items.ord_trait().unwrap();
+        self.cmp_method_ref(ord_def_id, "le", ty)
+    }
+
+    fn num_variants(&mut self, adt_def: ty::AdtDef<'tcx>) -> usize {
+        adt_def.variants.len()
+    }
+
+    fn fields(&mut self, adt_def: ty::AdtDef<'tcx>, variant_index: usize) -> Vec<Field<Self>> {
+        adt_def.variants[variant_index]
+            .fields
+            .iter()
+            .enumerate()
+            .map(|(index, field)| {
+                if field.name == special_idents::unnamed_field.name {
+                    Field::Indexed(index)
+                } else {
+                    Field::Named(field.name)
+                }
+            })
+            .collect()
+    }
+
+    fn needs_drop(&mut self, ty: Ty<'tcx>, span: Self::Span) -> bool {
+        if self.infcx.type_moves_by_default(ty, span) {
+            // FIXME(#21859) we should do an add'l check here to determine if
+            // any dtor will execute, but the relevant fn
+            // (`type_needs_drop`) is currently factored into
+            // `librustc_trans`, so we can't easily do so.
+            true
+        } else {
+            // if type implements Copy, cannot require drop
+            false
+        }
+    }
+
+    fn span_bug(&mut self, span: Self::Span, message: &str) -> ! {
+        self.tcx.sess.span_bug(span, message)
+    }
+}
+
+impl<'a,'tcx:'a> Cx<'a,'tcx> {
+    fn cmp_method_ref(&mut self,
+                      trait_def_id: DefId,
+                      method_name: &str,
+                      arg_ty: Ty<'tcx>)
+                      -> ItemRef<Cx<'a,'tcx>> {
+        let method_name = token::intern(method_name);
+        let substs = Substs::new_trait(vec![arg_ty], vec![], arg_ty);
+        for trait_item in self.tcx.trait_items(trait_def_id).iter() {
+            match *trait_item {
+                ty::ImplOrTraitItem::MethodTraitItem(ref method) => {
+                    if method.name == method_name {
+                        let method_ty = self.tcx.lookup_item_type(method.def_id);
+                        let method_ty = method_ty.ty.subst(self.tcx, &substs);
+                        return ItemRef {
+                            ty: method_ty,
+                            def_id: method.def_id,
+                            substs: self.tcx.mk_substs(substs),
+                        };
+                    }
+                }
+                ty::ImplOrTraitItem::ConstTraitItem(..) |
+                ty::ImplOrTraitItem::TypeTraitItem(..) => {
+                }
+            }
+        }
+
+        self.tcx.sess.bug(
+            &format!("found no method `{}` in `{:?}`", method_name, trait_def_id));
+    }
+}
+
+impl<'a,'tcx> PartialEq for Cx<'a,'tcx> {
+    fn eq(&self, _other: &Cx<'a,'tcx>) -> bool {
+        false
```

Comment:
`PartialEq` does not guarantee reflexivity, so always returning `false` is actually totally fine: <https://doc.rust-lang.org/beta/std/cmp/trait.PartialEq.html>